crates/neon/src/instance/mod.rs

```diff
@@ -0,0 +1,72 @@
+use std::any::Any;
+use std::marker::PhantomData;
+use std::sync::atomic::{AtomicUsize, Ordering};
+
+use once_cell::sync::OnceCell;
+
+use crate::context::Context;
+use crate::lifecycle::InstanceData;
+
+static COUNTER: AtomicUsize = AtomicUsize::new(0);
+
+fn next_id() -> usize {
+    COUNTER.fetch_add(1, Ordering::SeqCst)
+}
+
+/// A cell that can be used to allocate data that is global to an instance
+/// of a Neon addon.
+pub struct Global<T> {
+    _type: PhantomData<T>,
+    id: OnceCell<usize>,
+}
+
+impl<T> Global<T> {
+    pub const fn new() -> Self {
+        Self {
+            _type: PhantomData,
+            id: OnceCell::new(),
+        }
+    }
+
+    fn id(&self) -> usize {
+        *self.id.get_or_init(next_id)
+    }
+}
+
+impl<T: Any + Send> Global<T> {
+    pub fn borrow<'a, 'b, C>(&self, cx: &'b mut C) -> Option<&'b T>
+    where
+        C: Context<'a>,
+    {
+        InstanceData::globals(cx)[self.id()]
+            .as_ref()
+            .map(|boxed| boxed.downcast_ref().unwrap())
+    }
+
+    pub fn borrow_mut<'a, 'b, C>(&self, cx: &'b mut C) -> Option<&'b mut T>
+    where
+        C: Context<'a>,
+    {
+        InstanceData::globals(cx)[self.id()]
+            .as_mut()
+            .map(|boxed| boxed.downcast_mut().unwrap())
+    }
+
+    pub fn set<'a, 'b, C>(&self, cx: &'b mut C, v: T)
+    where
+        C: Context<'a>,
+    {
+        InstanceData::globals(cx)[self.id()] = Some(Box::new(v));
+    }
+}
+
+impl<T: Any + Send + Clone> Global<T> {
+    pub fn get<'a, C>(&self, cx: &mut C) -> Option<T>
```

Comment:
A `get` that clones is a little strange. I think this should be removed since it could be replaced without much boilerplate:

```rust
// With `get`
let data = DATA.get(&cx);

// Without `get`
let data = DATA.borrow(&cx).cloned();
```

If we do keep this, I think it shoudl take `&C` and not `&mut C`.

Reply:
Ooh, I didn't know about `Option::cloned()`, that does help. I ran into annoying paper cuts with `get` because it keeps the borrow of the context live, so you can easily end up with borrowck errors like this:

![image](https://user-images.githubusercontent.com/307871/170142785-f3229ec4-db48-4a35-8c98-2a1ded09acf0.png)

But that's pretty easily written like this:

```rust
pub fn get_thread_id(mut cx: FunctionContext) -> JsResult<JsValue> {
    let id = THREAD_ID.get(&mut cx).cloned();
    Ok(match id {
        Some(id) => cx.number(id).upcast(),
        None => cx.undefined().upcast(),
    })
}
```

I'll drop the method for now and maybe recommend `Option::cloned()` in the API docs for `get()` as a useful idiom.

Reply:
I think this might have been the reason we originally considered a `OnceCell` model that *doesn't* give mutability.

If we never vend a `&mut T`, we could give `&'cx T` instead of `&'a T`. I'm still not sure which is better; requiring a reference counting/cloning for *this* use case or requiring interior mutability for changing.

Reply:
I just pushed a commit that eliminates all the `&mut T` and uses `&'cx T` instead of `&'a T`. Let's take a look and see what we think!